src/main/java/org/benetech/mathshare/service/impl/ProblemSetServiceImpl.java

```diff
@@ -52,13 +52,21 @@ public ProblemSetRevision getProblemSetByShareUrl(long shareUrl) {
     }
 
     @Override
-    public ProblemSetRevision saveNewProblemSet(ProblemSet problemSet) throws IllegalArgumentException {
-        if (problemSet.getId() != null) {
-            throw new IllegalArgumentException("Id must be null for the new problem set");
+    public ProblemSetRevision saveNewProblemSet(ProblemSetDTO problemSetDTO) throws IllegalArgumentException {
+        ProblemSet problemSet = ProblemMapper.INSTANCE.fromDto(problemSetDTO);
+        List<Problem> problems = problemSetDTO.getProblems().stream().map(ProblemMapper.INSTANCE::fromDto)
+                .collect(Collectors.toList());
+        List<Problem> savedProblems = new ArrayList<>();
+
+        ProblemSet set = problemSetRepository.save(problemSet);
+        em.refresh(set);
+        ProblemSetRevision revision = problemSetRevisionRepository.save(new ProblemSetRevision(set));
+        for (Problem problem : problems) {
+            savedProblems.add(createOrUpdateProblem(problem, revision));
         }
-        ProblemSet savedProblemSet = problemSetRepository.save(problemSet);
-        return problemSetRevisionRepository.save(
-                new ProblemSetRevision(savedProblemSet));
+        revision.setProblems(savedProblems);
+        em.refresh(revision);
```

Comment:
There's probably no need to refresh not-saved entity

Reply:
it is being saved